Source/Csla.Blazor/State/StateManager.cs

```diff
@@ -0,0 +1,129 @@
+//     Website: https://cslanet.com
+// </copyright>
+// <summary>Get and save state from Blazor pages</summary>
+//-----------------------------------------------------------------------
+using System;
+using System.Threading.Tasks;
+using Csla.State;
+
+namespace Csla.Blazor.State
+{
+  /// <summary>
+  /// Get and save state from Blazor pages.
+  /// </summary>
+  /// <param name="sessionManager"></param>
+  public class StateManager(ISessionManager sessionManager)
+  {
+    private readonly ISessionManager _sessionManager = sessionManager;
+
+    /// <summary>
+    /// Get state from web server for use in Blazor wasm. Must call
+    /// as user navigates from any server-side Blazor page to a wasm
+    /// Blazor page.
+    /// client.
+    /// </summary>
+    /// <param name="calledFrom">Location method is invoked in page</param>
+    /// <param name="firstRender">firstRender value from OnAfterRender method</param>
+    /// <returns></returns>
+    /// <remarks>
+    /// Normally this is called in both the initialize and after render
+    /// methods of a Blazor page. This way the page can be flexibly
+    /// rendered in server-rendered, server-interactive, and wasm-interactive
+    /// modes and the state will always be available.
+    /// </remarks>
+    public async Task GetState(CalledFrom calledFrom, bool firstRender)
+    {
+      await GetState(calledFrom, firstRender, 10);
+    }
+
+    /// <summary>
+    /// Get state from web server for use in Blazor wasm. Must call
+    /// as user navigates from any server-side Blazor page to a wasm
+    /// Blazor page.
+    /// client.
+    /// </summary>
+    /// <param name="calledFrom">Location method is invoked in page</param>
+    /// <param name="firstRender">firstRender value from OnAfterRender method</param>
+    /// <param name="timeout">Milliseconds to wait before timing out</param>
+    /// <returns></returns>
+    /// <remarks>
+    /// Normally this is called in both the initialize and after render
+    /// methods of a Blazor page. This way the page can be flexibly
+    /// rendered in server-rendered, server-interactive, and wasm-interactive
+    /// modes and the state will always be available.
+    /// </remarks>
+    public async Task GetState(CalledFrom calledFrom, bool firstRender, int timeout)
+    {
+      var isBrowser = OperatingSystem.IsBrowser();
+      if (isBrowser)
+      {
+        if (calledFrom == CalledFrom.AfterRender && firstRender)
+        {
+            await _sessionManager.RetrieveSession();
+        }
+      }
+      else
+      {
+        if (calledFrom == CalledFrom.Initialize)
+        {
+          await WaitForState(timeout);
+        }
+      }
+    }
+
+    /// <summary>
+    /// Waits for state to be available as it is transferred
+    /// from Blazor wasm to the web server.
+    /// </summary>
+    /// <param name="timeout">Milliseconds to wait before timing out</param>
+    private async Task WaitForState(int timeout)
```

Comment:
The description does not match the actual implemention. (IMHO I would prefer a TimeSpan over an int to make the intent more clear and no need to know in which unit of time it's actually used).
Timeout is used as a counter and the delay is always 1s.

Tbh: I would rewrite this to use a `TaskCompletionSource` and either the Timeout kicks in or the sessions `ICheckedOut` changes. IIRC the session provides a property changed event, no?
So you can do something like this:
```csharp
var  session = _sessionManager.GetSession();
var tcs = new TaskCompletionSource();
try
{
    session.PropertyChanged += Handler;
    await tcs.Task.WaitAsync(<TimeOutAsTimeSpan>); // throws TimeoutException when Timeout hit
}
finally
{
    session.PropertyChanged -= Handler;
}

void Handler(<CorrectParamsForPropertyChangedHere>)
{
    switch(e.PropertyName)
    {
        case nameof(session.IsCheckedOut):
            if(!session.IsCheckedOut)
            {
                tcs.SetResult();
            }
            break;
    }
}
```

Reply:
Hmm. I tried this, but it blocks the thread, preventing the wasm page from disposing, and so results in a deadlock. The spinlock I used by waiting for `IsCheckedOut` to change doesn't block the thread, and therefore allows the wasm page to dispose - which results in the call to the controller to update the server state and unlock `Session`.

Reply:
Oh, sure, now my _original_ implementation is locking too. Damn!